core/zoe/contracts/simpleOffer/srcs/secondPriceSrcs.js

```diff
@@ -0,0 +1,107 @@
+import harden from '@agoric/harden';
+
+import { makeHasOkRules } from '../../../contractUtils';
+
+const hasOkRulesInitialOffer = makeHasOkRules([
+  ['haveExactly', 'wantAtLeast'], // The initial offer
+]);
+
+const hasOkRulesBids = makeHasOkRules([
+  ['wantExactly', 'haveAtMost'], // Subsequent bids
+]);
+
+const isValidInitialOfferDesc = newOfferDesc =>
+  hasOkRulesInitialOffer(newOfferDesc);
+
+const isValidBid = (assays, initialOffer, newBid) =>
+  hasOkRulesBids(newBid) &&
+  assays[1].includes(newBid[1].amount, initialOffer[1].amount);
+
+const isValidOffer = (
+  assays,
+  offerIds,
+  offerIdsToOfferDescs,
+  offerMadeDesc,
+) => {
+  const isInitialOffer = offerIds.length === 0;
+  return (
+    (isInitialOffer && isValidInitialOfferDesc(offerMadeDesc)) ||
+    (!isInitialOffer &&
+      isValidBid(assays, offerIdsToOfferDescs.get(offerIds[0]), offerMadeDesc))
+  );
+};
+
+// Iterate over all of the bids keeping the highest and second highest bid.
+const findWinnerAndPrice = (strategy, offerIds, bids) => {
+  let highestBid = strategy.empty();
+  let secondHighestBid = strategy.empty();
+  let offerIdHighestBid;
+  // If the bid is greater than the highest bid, it is the new highest
+  // bid.
+  // Has side effects
+  // eslint-disable-next-line array-callback-return
+  offerIds.map((offerId, i) => {
+    const bid = bids[i];
+    if (strategy.includes(bid, highestBid)) {
+      secondHighestBid = highestBid;
+      highestBid = bid;
+      offerIdHighestBid = offerId;
+    } else if (strategy.includes(bid, secondHighestBid)) {
+      // If the bid is not greater than the highest bid, but is greater
+      // than the second highest bid, it is the new second highest bid.
+      secondHighestBid = bid;
+    }
+  });
+  return {
+    winner: offerIdHighestBid,
+    price: secondHighestBid,
+  };
+};
+
+const reallocate = (
+  strategies,
+  offerIds,
+  offerIdsToOfferDescs,
+  getQuantitiesFor,
+) => {
+  const bids = offerIds
+    .map(offerId => offerIdsToOfferDescs.get(offerId)[1].amount.quantity)
+    .slice(1);
+  const { winner, price } = findWinnerAndPrice(
+    strategies[1],
+    offerIds.slice(1),
+    bids,
+  );
+
+  const creator = offerIds[0];
+  const creatorQuantities = getQuantitiesFor(harden([creator]))[0];
```

Comment:
All (nearly all? I can't tell) of these zeroes and ones should be replaced with manifest constants indicating that 0 is the itemIndex, and 1 is the priceIndex.  Above, there are cases where 0 is the sellerIndex (and some 1s are really `sellerIndex + 1`. I'm not sure what other 0s and 1s might be in use, but it'll be easier to see once these have names.

Reply:
Yeah, there are multiple different reasons for indexing into an array with zero. You're right that it's confusing. Let me see what I can do

Reply:
Ok I think I pushed up those changes. Let me know what you think